specs/booking_spec.rb

```diff
@@ -0,0 +1,428 @@
+    end # it "should not reserve the same room more than once for a date range" do
+
+    it "will add sequential rooms to the @all_reservation array" do
+      @booking.make_reservation(@checkin_day, @checkout_day, 1)
+      @booking.make_reservation(@checkin_day, @checkout_day, 1)
+      @booking.make_reservation(@checkin_day, @checkout_day, 3)
+
+      # room_array = []
+
+      @booking.all_reservations[0].res_rooms.must_equal [@booking.all_rooms[0]]
+      @booking.all_reservations[1].res_rooms.must_equal [@booking.all_rooms[1]]
+      @booking.all_reservations[2].res_rooms[0].must_equal @booking.all_rooms[2]
+      @booking.all_reservations[2].res_rooms[1].must_equal @booking.all_rooms[3]
+      @booking.all_reservations[2].res_rooms[2].must_equal @booking.all_rooms[4]
+
+    end # it "will add sequential rooms to the availe array" do
+
+    it "will never include the same room twice during a given date range" do
+      @booking.make_reservation(@checkin_day, @checkout_day, 1)
+      @booking.make_reservation(@checkin_day, @checkout_day, 1)
+      @booking.make_reservation(@checkin_day, @checkout_day, 3)
+
+      rooms = []
+      @booking.check_date_for_reservations(@checkin_day, @checkout_day)
+      @booking.check_date_for_reservations(@checkin_day, @checkout_day).each do |res|
+        res.res_rooms.each do |room|
+          rooms << room
+        end #.each
+      end # .each
+
+      rooms.length.must_equal 5
+      rooms.detect{ |room| rooms.count(room) > 1}.must_equal nil
+    end # it "will never include the same room twice during a given date range" do
+
+    it "will raise BookingError if more rooms are asked for then the hotel has (>20)" do
+      proc{@booking.make_reservation(@checkin_day, @checkout_day, 21)}.must_raise BookingError
+    end # it "will raise BookingError if more than 20 rooms are asked for
+
+    it "will raise BookingError if more rooms are requested then are availible for that date range" do
+      @booking.make_reservation(@checkin_day, @checkout_day, 5)
+      proc{@booking.make_reservation(@checkin_day, @checkout_day, 18)}.must_raise BookingError
+    end # it "will raise BookingError if more rooms are requested then are availible
+
+    it "will be able to book a room the day that it opens up " do
+       @booking.make_reservation(@checkin_day, @checkout_day, 20)
+      #  @booking.make_reservation(@checkout_day, (@checkout_day + 2), 20).must_be_kind_of Array
+       @booking.make_reservation(@checkout_day, (@checkout_day + 2), 20)[0].res_rooms.length.must_equal 20
+    end # it "will be able to book a room the day that it opens up " do
```

Comment:
These tests are a good start, but there are many more interesting cases around whether a room is available that I would like to see. My entire list includes:
- Not available:
  - Same dates
  - Overlaps in the front
  - Overlaps in the back
  - Completely contained
  - Completely containing
- Available:
  - Completely before
  - Completely after
  - Ends on the checkin date
  - Starts on the checkout date